Intermittent build failures in `BookKeeperLogTests > testAppendTransientZooKeeperFailure`

**Problem description**
io.pravega.segmentstore.storage.impl.bookkeeper.BookKeeperLogTests > testAppendTransientZooKeeperFailure FAILED
    java.lang.AssertionError: Unexpected sequence number. expected:<51539607561> but was:<55834574848>
        at org.junit.Assert.fail(Assert.java:88)
        at org.junit.Assert.failNotEquals(Assert.java:834)
        at org.junit.Assert.assertEquals(Assert.java:645)
        at 

**Problem location**
Unittests for `BookKeeperLogTests > testAppendTransientZooKeeperFailure`
**Suggestions for an improvement**

--- a/segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/LogReader.java
+++ b/segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/LogReader.java
@@ -118,9 +118,19 @@ class LogReader implements CloseableIterator<DurableDataLog.ReadItem, DurableDat
 
         LedgerMetadata metadata = this.metadata.getLedger(address.getLedgerId());
         assert metadata != null : "no LedgerMetadata could be found with valid LedgerAddress " + address;
+        val allMetadatas = this.metadata.getLedgers();
 
         // Open the ledger.
-        val ledger = Ledgers.openRead(metadata.getLedgerId(), this.bookKeeper, this.config);
+        LedgerHandle ledger;
+        if (allMetadatas.size() == 0 || metadata == allMetadatas.get(allMetadatas.size() - 1)) {
+            // This is our last ledger (the active one); we need to make sure open it without recovery since otherwise we
+            // we would fence ourselves out.
+            ledger = Ledgers.openRead(metadata.getLedgerId(), this.bookKeeper, this.config);
+        } else {
+            // Older ledger. Open with recovery to make sure any uncommitted fragments will be recovered. Since we do our
+            // Log fencing based on the last Ledger, open-fencing this Ledger will not have any adverse effects.
+            ledger = Ledgers.openFence(metadata.getLedgerId(), this.bookKeeper, this.config);
+        }
 
         long lastEntryId = ledger.getLastAddConfirmed();
         if (lastEntryId < address.getEntryId()) {
@@ -130,9 +140,15 @@ class LogReader implements CloseableIterator<DurableDataLog.ReadItem, DurableDat
             return;
         }
 
+        ReadLedger previousLedger;
         try {
             val reader = Exceptions.handleInterrupted(() -> ledger.readEntries(address.getEntryId(), lastEntryId));
+            previousLedger = this.currentLedger;
             this.currentLedger = new ReadLedger(metadata, ledger, reader);
+            if (previousLedger != null) {
+                // Close previous ledger handle.
+                Ledgers.close(previousLedger.handle);
+            }
         } catch (Exception ex) {
             Ledgers.close(ledger);
             close();

--- a/segmentstore/storage/src/test/java/io/pravega/segmentstore/storage/DurableDataLogTestBase.java
+++ b/segmentstore/storage/src/test/java/io/pravega/segmentstore/storage/DurableDataLogTestBase.java
@@ -129,6 +129,31 @@ public abstract class DurableDataLogTestBase extends ThreadPooledTestSuite {
     }
 
     /**
+     * Tests the ability to execute reads after recovery, immediately followed by writes.
+     *
+     * @throws Exception If one got thrown.
+     */
+    @Test
+    public void testReadWriteRecovery() throws Exception {
+        final int iterationCount = 4;
+        Object context = createSharedContext();
+        TreeMap<LogAddress, byte[]> writeData = new TreeMap<>(Comparator.comparingLong(LogAddress::getSequence));
+        for (int i = 0; i < iterationCount; i++) {
+            try (DurableDataLog log = createDurableDataLog(context)) {
+                log.initialize(TIMEOUT);
+                verifyReads(log, writeData);
+                writeData.putAll(populate(log, getWriteCount()));
+            }
+        }
+
+        // One last recovery at the end.
+        try (DurableDataLog log = createDurableDataLog(context)) {
+            log.initialize(TIMEOUT);
+            verifyReads(log, writeData);
+        }
+    }
+
+    /**
      * Tests the ability to truncate from a DurableDataLog.
      *
      * @throws Exception If one got thrown.

